# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGLProgram.h
# Commit: ccfa767dba64
# Full Hash: ccfa767dba644dc193e0246eb8e8ff3377e8b8a5
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Client-side bindings mirror for precise CC, and merge similar codepaths. r=handyman
#   
#   * Context loss using RAII
#   * Move Program reflection Client-side
#   
# ==============================================================================

diff -r 71c122ac0ca7 -r ccfa767dba64 dom/canvas/WebGLProgram.h
--- a/dom/canvas/WebGLProgram.h	Wed Jan 08 22:19:14 2020 +0000
+++ b/dom/canvas/WebGLProgram.h	Wed Jan 08 22:19:16 2020 +0000
@@ -8,27 +8,20 @@
 
 #include <map>
 #include <set>
-#include <string>
 #include <vector>
 
-#include "mozilla/LinkedList.h"
 #include "mozilla/RefPtr.h"
 #include "mozilla/WeakPtr.h"
-#include "nsString.h"
-#include "nsWrapperCache.h"
 
 #include "CacheInvalidator.h"
-#include "WebGLActiveInfo.h"
 #include "WebGLContext.h"
 #include "WebGLObjectModel.h"
 
 namespace mozilla {
 class ErrorResult;
-class WebGLActiveInfo;
 class WebGLContext;
 class WebGLProgram;
 class WebGLShader;
-class WebGLUniformLocation;
 
 namespace dom {
 template <typename>
@@ -42,48 +35,15 @@
 
 enum class TextureBaseType : uint8_t;
 
-struct AttribInfo final {
-  const WebGLActiveInfo mActiveInfo;
-  const GLint mLoc;  // -1 for active built-ins
-};
-
-struct UniformInfo final {
-  typedef decltype(WebGLContext::mBound2DTextures) TexListT;
-
-  const WebGLActiveInfo mActiveInfo;
-  const TexListT* const mSamplerTexList;
-  const webgl::TextureBaseType mTexBaseType;
-  const bool mIsShadowSampler;
-
-  std::vector<uint32_t> mSamplerValues;
-
- protected:
-  static const TexListT* GetTexList(const WebGLContext* aWebGL,
-                                    WebGLActiveInfo* activeInfo);
-
- public:
-  explicit UniformInfo(const WebGLContext* aWebGL, WebGLActiveInfo& activeInfo);
-};
-
 struct UniformBlockInfo final {
-  const nsCString mUserName;
-  const nsCString mMappedName;
-  const uint32_t mDataSize;
-
-  const IndexedBufferBinding* mBinding;
-
-  UniformBlockInfo(WebGLContext* webgl, const nsACString& userName,
-                   const nsACString& mappedName, uint32_t dataSize)
-      : mUserName(userName),
-        mMappedName(mappedName),
-        mDataSize(dataSize),
-        mBinding(&webgl->mIndexedUniformBufferBindings[0]) {}
+  const ActiveUniformBlockInfo& info;
+  const IndexedBufferBinding* binding = nullptr;
 };
 
 struct FragOutputInfo final {
   const uint8_t loc;
-  const nsCString userName;
-  const nsCString mappedName;
+  const std::string userName;
+  const std::string mappedName;
   const TextureBaseType baseType;
 };
 
@@ -93,6 +53,34 @@
   std::vector<BufferAndIndex> usedBuffers;
 };
 
+// -
+
+void UniformAs1fv(gl::GLContext& gl, GLint location, GLsizei count,
+                  bool transpose, const void* any);
+
+struct ActiveUniformValidationInfo final {
+  const ActiveUniformInfo& info;
+  uint8_t channelsPerElem = 0;
+  decltype(&UniformAs1fv) pfn = nullptr;
+
+  static ActiveUniformValidationInfo Make(const ActiveUniformInfo&);
+};
+
+struct SamplerUniformInfo final {
+  const decltype(WebGLContext::mBound2DTextures)& texListForType;
+  const webgl::TextureBaseType texBaseType;
+  const bool isShadowSampler;
+  std::vector<uint32_t> texUnits;
+};
+
+struct LocationInfo final {
+  const ActiveUniformValidationInfo info;
+  const uint32_t indexIntoUniform;
+  SamplerUniformInfo* const samplerInfo;
+};
+
+// -
+
 struct LinkedProgramInfo final : public RefCounted<LinkedProgramInfo>,
                                  public SupportsWeakPtr<LinkedProgramInfo>,
                                  public CacheInvalidator {
@@ -106,20 +94,23 @@
   WebGLProgram* const prog;
   const GLenum transformFeedbackBufferMode;
 
-  std::vector<AttribInfo> attribs;
-  std::vector<UniformInfo*> uniforms;            // Owns its contents.
-  std::vector<UniformBlockInfo*> uniformBlocks;  // Owns its contents.
-  std::vector<WebGLActiveInfo> transformFeedbackVaryings;
   std::unordered_map<uint8_t, const FragOutputInfo> fragOutputs;
   uint8_t zLayerCount = 1;
 
-  // Needed for draw call validation.
-  std::vector<UniformInfo*> uniformSamplers;
-
   mutable std::vector<size_t> componentsPerTFVert;
 
   bool attrib0Active;
 
+  // -
+
+  std::map<std::string, std::string> nameMap;
+  webgl::LinkActiveInfo active;
+
+  std::vector<std::unique_ptr<SamplerUniformInfo>> samplerUniforms;
+  std::unordered_map<uint32_t, LocationInfo> locationMap;
+
+  mutable std::vector<UniformBlockInfo> uniformBlocks;
+
   //////
 
   mutable CacheWeakMap<const WebGLVertexArray*, CachedDrawFetchLimits>
@@ -131,49 +122,25 @@
 
   explicit LinkedProgramInfo(WebGLProgram* prog);
   ~LinkedProgramInfo();
-
-  bool FindAttrib(const nsCString& userName,
-                  const AttribInfo** const out_info) const;
-  bool FindUniform(const nsCString& userName, nsCString* const out_mappedName,
-                   size_t* const out_arrayIndex,
-                   UniformInfo** const out_info) const;
 };
 
 }  // namespace webgl
 
-class WebGLProgram final : public WebGLRefCountedObject<WebGLProgram>,
-                           public LinkedListElement<WebGLProgram> {
+class WebGLProgram final : public WebGLContextBoundObject {
   friend class WebGLTransformFeedback;
   friend struct webgl::LinkedProgramInfo;
 
+  MOZ_DECLARE_REFCOUNTED_VIRTUAL_TYPENAME(WebGLProgram, override)
+
  public:
-  NS_INLINE_DECL_REFCOUNTING(WebGLProgram)
-
   explicit WebGLProgram(WebGLContext* webgl);
 
   void Delete();
 
   // GL funcs
-  void AttachShader(WebGLShader* shader);
-  void BindAttribLocation(GLuint index, const nsAString& name);
-  void DetachShader(const WebGLShader* shader);
-  Maybe<WebGLActiveInfo> GetActiveAttrib(GLuint index) const;
-  Maybe<WebGLActiveInfo> GetActiveUniform(GLuint index) const;
-  MaybeAttachedShaders GetAttachedShaders() const;
-  GLint GetAttribLocation(const nsAString& name) const;
-  GLint GetFragDataLocation(const nsAString& name) const;
-  nsString GetProgramInfoLog() const;
-  MaybeWebGLVariant GetProgramParameter(GLenum pname) const;
-  GLuint GetUniformBlockIndex(const nsAString& name) const;
-  nsString GetActiveUniformBlockName(GLuint uniformBlockIndex) const;
-  MaybeWebGLVariant GetActiveUniformBlockParam(GLuint uniformBlockIndex,
-                                               GLenum pname) const;
-  MaybeWebGLVariant GetActiveUniformBlockActiveUniforms(
-      GLuint uniformBlockIndex) const;
-  already_AddRefed<WebGLUniformLocation> GetUniformLocation(
-      const nsAString& name) const;
-  MaybeWebGLVariant GetUniformIndices(
-      const nsTArray<nsString>& uniformNames) const;
+  void AttachShader(WebGLShader& shader);
+  void BindAttribLocation(GLuint index, const std::string& name);
+  void DetachShader(const WebGLShader& shader);
   void UniformBlockBinding(GLuint uniformBlockIndex,
                            GLuint uniformBlockBinding) const;
 
@@ -183,23 +150,8 @@
 
   ////////////////
 
-  bool FindAttribUserNameByMappedName(const nsACString& mappedName,
-                                      nsCString* const out_userName) const;
-  bool FindVaryingByMappedName(const nsACString& mappedName,
-                               nsCString* const out_userName,
-                               bool* const out_isArray) const;
-  bool FindUniformByMappedName(const nsACString& mappedName,
-                               nsCString* const out_userName,
-                               bool* const out_isArray) const;
-  bool UnmapUniformBlockName(const nsCString& mappedName,
-                             nsCString* const out_userName) const;
-
-  void TransformFeedbackVaryings(const nsTArray<nsString>& varyings,
+  void TransformFeedbackVaryings(const std::vector<std::string>& varyings,
                                  GLenum bufferMode);
-  Maybe<WebGLActiveInfo> GetTransformFeedbackVarying(GLuint index) const;
-
-  void EnumerateFragOutputs(
-      std::map<nsCString, const nsCString>& out_FragOutputs) const;
 
   bool IsLinked() const { return mMostRecentLinkInfo; }
 
@@ -210,27 +162,29 @@
   const auto& VertShader() const { return mVertShader; }
   const auto& FragShader() const { return mFragShader; }
 
+  const auto& LinkLog() const { return mLinkLog; }
+
  private:
   ~WebGLProgram();
 
   void LinkAndUpdate();
   bool ValidateForLink();
-  bool ValidateAfterTentativeLink(nsCString* const out_linkLog) const;
+  bool ValidateAfterTentativeLink(std::string* const out_linkLog) const;
 
  public:
   const GLuint mGLName;
 
  private:
-  WebGLRefPtr<WebGLShader> mVertShader;
-  WebGLRefPtr<WebGLShader> mFragShader;
+  RefPtr<WebGLShader> mVertShader;
+  RefPtr<WebGLShader> mFragShader;
   size_t mNumActiveTFOs;
 
   std::map<std::string, GLuint> mNextLink_BoundAttribLocs;
 
-  std::vector<nsString> mNextLink_TransformFeedbackVaryings;
+  std::vector<std::string> mNextLink_TransformFeedbackVaryings;
   GLenum mNextLink_TransformFeedbackBufferMode;
 
-  nsCString mLinkLog;
+  std::string mLinkLog;
   RefPtr<const webgl::LinkedProgramInfo> mMostRecentLinkInfo;
 };
 